plutus-use-cases/src/Language/PlutusTx/Coordination/Contracts/Vesting.hs

```diff
@@ -87,56 +97,102 @@ instance Eq VestingData where
 
 PlutusTx.makeLift ''VestingData
 
+type VestingSchema =
+    BlockchainActions
+        .\/ Endpoint "vest funds" ()
+        .\/ Endpoint "retrieve funds" Value
+
+-- | The current state of the vesting contract
+data VestingState =
+    VestingState
+        { knownOutputs :: AddressMap
+        , vestingData  :: VestingData
+        }
+
+initialState :: Vesting -> VestingState
+initialState vd =
+    let hash = validatorScriptHash vd
+        addr = Ledger.scriptAddress $ validatorScript vd
+    in  VestingState
+            { knownOutputs = AM.addAddress addr Prelude.mempty
+            , vestingData  = VestingData hash mempty
+            }
+
+type MonadVesting m =
+    ( MonadState VestingState m 
+    , MonadReader Vesting m
+    , MonadBase (Contract VestingSchema) m
+    )
+
+contractAddress :: MonadReader Vesting m => m Address
+contractAddress = asks (Ledger.scriptAddress . validatorScript)
+    
+-- | Lock some Ada in the contract using the 'VestingData' as data script.
+payIntoContract
+    :: MonadReader Vesting m
+    => VestingData
+    -> Value
+    -> m TxOut
+payIntoContract vd vl = do
+    address <- contractAddress
+    let datScript = DataScript $ Ledger.lifted vd
+    pure $ LTx.scriptTxOut' vl address datScript
+
+
+vestFundsC
+    :: ( MonadVesting m )
+    => m ()
+vestFundsC = do
+    () <- liftBase (endpoint @"vest funds")
+    h <- asks validatorScriptHash
+    total <- asks totalAmount
+    out <- payIntoContract (VestingData h mempty) total
+    liftBase $ writeTx $ Tx.unbalancedTx [] [out]
+
+
+retrieveFundsC
+    :: ( MonadVesting m )
+    => m ()
+retrieveFundsC = do
+    vl <- liftBase (endpoint  @"retrieve funds")
+    currentSlot <- liftBase (awaitSlot 0)
+    total <- asks totalAmount
+    valScript <- asks validatorScript
+    VestingState utxo vst <- get
+    let ins = AM.spendScriptOutputs valScript redeemerScript utxo
+        newState = vst { vestingDataPaidOut = vl + vestingDataPaidOut vst }
+    outs <- payIntoContract newState (total - vestingDataPaidOut newState)
+    liftBase $ writeTx $ Tx.unbalancedTx ins [outs] & Tx.validityRange .~ Interval.from currentSlot
```

Comment:
So this always creates a new locked output even if we've retrieved all the funds?

Reply:
This is not the case anymore in the new version